component-runtime-beam/src/main/java/org/talend/sdk/component/runtime/beam/spi/BeamProducerFinder.java

```diff
@@ -111,12 +125,20 @@ public Record next() {
         }
 
         private Record findNext() {
-            Record record = QUEUE.poll();
-
+            System.out.println("Find Next, Thread " + Thread.currentThread().getId() +"; Queue size " + QUEUE.size());
+            System.out.flush();
+            Record record = QUEUE.get(this.queueIndex).poll();
+            int index = 0;
             while (record == null && (!end)) {
-                end = result.getState() != PipelineResult.State.RUNNING;
+                end = result != null && result.getState() != PipelineResult.State.RUNNING;
+                if (!end && index > 10) {
```

Comment:
I don't get the value _10_ here...

Reply:
here, main thread is waiting that beam thread put element in the QUEUE (with a sleep between each try).
After 10 try (can be parameterized, as the duration of sleep), we wait the full beam process to terminate.